File: src/main/java/com/chatchatabc/jpademojava/application/rest/CountryController.java
package com.chatchatabc.jpademojava.application.rest;

import com.chatchatabc.jpademojava.application.dto.ErrorContent;
import com.chatchatabc.jpademojava.application.dto.country.*;
import com.chatchatabc.jpademojava.domain.model.Country;
import com.chatchatabc.jpademojava.domain.model.User;
import com.chatchatabc.jpademojava.domain.repository.CountryRepository;
import com.chatchatabc.jpademojava.domain.repository.UserRepository;
import com.chatchatabc.jpademojava.domain.service.CountryService;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;

@RestController
@RequestMapping("/api/country")
public class CountryController {
    @Autowired
    private CountryService countryService;
    @Autowired
    private CountryRepository countryRepository;
    @Autowired
    private UserRepository userRepository;

    private final ModelMapper mapper = new ModelMapper();

    /**
     * Get countries
     *
     * @param pageable
     * @return
     */
    @GetMapping("/get")
    public ResponseEntity<Page<Country>> getCountries(
            Pageable pageable
    ) {
        try {
            Page<Country> countries = countryRepository.findAll(pageable);
            return ResponseEntity.ok(countries);
        } catch (Exception e) {
            return ResponseEntity.badRequest().build();
        }
    }

    /**
     * Get users by country
     *
     * @param countryId
     * @param pageable
     * @return
     */
    @GetMapping("/get/{countryId}")
    public ResponseEntity<Page<User>> getUsersByCountry(
            @PathVariable String countryId,
            Pageable pageable
    ) {
        try {
            Optional<Country> country = countryRepository.findById(countryId);
            if (country.isEmpty()) {
                throw new Exception("Country not found");
            }
            Page<User> users = userRepository.findUsersByCountry(country.get(), pageable);
            return ResponseEntity.ok(users);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.badRequest().build();
        }
    }

    /**
     * Create country
     *
     * @param request
     * @return
     */
    @PostMapping("/create")
    public ResponseEntity<CountryCreateResponse> createCountry(
            @RequestBody CountryCreateRequest request
    ) {
        try {
            Country newCountry = mapper.map(request, Country.class);
            Country country = countryService.create(newCountry);
            return ResponseEntity.ok(new CountryCreateResponse(country, null));
        } catch (Exception e) {
            // TODO: Improve error message
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
                    new CountryCreateResponse(null, new ErrorContent("Create Country Error", e.getMessage()))
            );
        }
    }

    /**
     * Update country
     *
     * @param countryId
     * @param request
     * @return
     */
    @PutMapping("/update/{countryId}")
    public ResponseEntity<CountryUpdateResponse> updateCountry(
            @PathVariable String countryId,
            @RequestBody CountryUpdateRequest request
    ) {
        try {
            Country newCountryInfo = mapper.map(request, Country.class);
            Country country = countryService.update(countryId, newCountryInfo);
            return ResponseEntity.ok(new CountryUpdateResponse(country, null));
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
                    new CountryUpdateResponse(null, new ErrorContent("Update Country Error", e.getMessage()))
            );
        }
    }

    @PutMapping("/assign")
    public ResponseEntity<CountryAssignResponse> assignCountry(
            @RequestBody CountryAssignRequest request
    ) {
        try {
            User user = countryService.assign(request.getUserId(), request.getCountryId());
            return ResponseEntity.ok(new CountryAssignResponse(user, null));
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
                    new CountryAssignResponse(null, new ErrorContent("Assign Country Error", e.getMessage()))
            );
        }
    }
}

File: src/main/kotlin/com/chatchatabc/jpademo/application/rest/CountryController.kt
package com.chatchatabc.jpademo.application.rest

import com.chatchatabc.jpademo.application.dto.ErrorContent
import com.chatchatabc.jpademo.application.dto.country.*
import com.chatchatabc.jpademo.domain.model.Country
import com.chatchatabc.jpademo.domain.model.User
import com.chatchatabc.jpademo.domain.repository.CountryRepository
import com.chatchatabc.jpademo.domain.repository.UserRepository
import com.chatchatabc.jpademo.domain.service.CountryService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/country")
class CountryController(
    private val countryRepository: CountryRepository,
    private val countryService: CountryService,
    private val userRepository: UserRepository
) {

    /**
     * Get Countries
     */
    @GetMapping("/get")
    fun getCountries(
        pageable: Pageable
    ): ResponseEntity<Page<Country>> {
        return try {
            val countries = countryRepository.findAll(pageable)
            ResponseEntity.ok(countries)
        } catch (e: Exception) {
            ResponseEntity.badRequest().build()
        }
    }

    /**
     * Get Users by Country
     */
    @GetMapping("/get/{countryId}")
    fun getUsersByCountry(
        @PathVariable countryId: String,
        pageable: Pageable
    ): ResponseEntity<Page<User>> {
        return try {
            val country = countryRepository.findById(countryId)
            if (country.isEmpty) {
                throw Exception("Country not found")
            }
            val users = userRepository.findUsersByCountry(country.get(), pageable)
            ResponseEntity.ok(users)
        } catch (e: Exception) {
            e.printStackTrace()
            ResponseEntity.badRequest().build()
        }
    }

    /**
     * Create Country
     */
    @PostMapping("/create")
    fun createCountry(
        @RequestBody request: CountryCreateRequest
    ): ResponseEntity<CountryCreateResponse> {
        return try {
            val country = countryService.create(request)
            ResponseEntity.ok(CountryCreateResponse(country, null))
        } catch (e: Exception) {
            ResponseEntity.badRequest()
                .body(CountryCreateResponse(null, ErrorContent("Country Create Error", e.message)))
        }
    }

    /**
     * Update Country
     */
    @PutMapping("/update/{countryId}")
    fun updateCountry(
        @PathVariable countryId: String,
        @RequestBody request: CountryUpdateRequest
    ): ResponseEntity<CountryUpdateResponse> {
        return try {
            val country = countryService.update(countryId, request)
            ResponseEntity.ok(CountryUpdateResponse(country, null))
        } catch (e: Exception) {
            ResponseEntity.badRequest()
                .body(CountryUpdateResponse(null, ErrorContent("Country Update Error", e.message)))
        }
    }

    /**
     * Assign Country
     */
    @PutMapping("/assign")
    fun assignCountry(
        @RequestBody request: CountryAssignRequest
    ): ResponseEntity<CountryAssignResponse> {
        return try {
            val user = countryService.assign(request.userId, request.countryId)
            ResponseEntity.ok(CountryAssignResponse(user, null))
        } catch (e: Exception) {
            ResponseEntity.badRequest()
                .body(CountryAssignResponse(null, ErrorContent("Country Update Error", e.message)))
        }
    }

    /**
     * Unassign Country
     */
    @PutMapping("/unassign")
    fun unassignCountry(
        @RequestBody request: CountryUnassignRequest
    ): ResponseEntity<CountryAssignResponse> {
        return try {
            val user = countryService.unassign(request.userId)
            ResponseEntity.ok(CountryAssignResponse(user, null))
        } catch (e: Exception) {
            ResponseEntity.badRequest()
                .body(CountryAssignResponse(null, ErrorContent("Country Update Error", e.message)))
        }
    }

    /**
     * Delete Country
     */
    @DeleteMapping("/delete/{countryId}")
    fun deleteCountry(
        @PathVariable countryId: String
    ): ResponseEntity<String> {
        return try {
            countryRepository.deleteById(countryId)
            ResponseEntity.ok(null)
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(null)
        }
    }
}